File: SpotifyForTeens/Models/ContentFilter.swift
//
//  ContentFilter.swift
//  SpotifyForTeens
//
//  Created on 2025-12-29
//

import Foundation

enum ContentRating: Int {
    case everyone = 0       // All ages
    case teen = 1          // 13+
    case mature = 2        // 17+
    case explicit = 3      // Explicit content
}

struct ContentFilter {
    var maxRating: ContentRating
    var blockExplicitContent: Bool
    var blockedKeywords: [String]
    var allowedGenres: [String]?
    
    static let `default` = ContentFilter(
        maxRating: .teen,
        blockExplicitContent: true,
        blockedKeywords: [],
        allowedGenres: nil
    )
    
    static let strict = ContentFilter(
        maxRating: .everyone,
        blockExplicitContent: true,
        blockedKeywords: ["drug", "violence", "explicit"],
        allowedGenres: ["pop", "rock", "indie", "electronic", "classical"]
    )
    
    func isContentAllowed(_ track: Track) -> Bool {
        // Block explicit content if setting is enabled
        if blockExplicitContent && track.isExplicit {
            return false
        }
        
        // Check for blocked keywords in track name
        let trackNameLower = track.name.lowercased()
        for keyword in blockedKeywords {
            if trackNameLower.contains(keyword.lowercased()) {
                return false
            }
        }
        
        // Check for blocked keywords in artist names
        let artistNameLower = track.artistNames.lowercased()
        for keyword in blockedKeywords {
            if artistNameLower.contains(keyword.lowercased()) {
                return false
            }
        }
        
        return true
    }
}

class ParentalControlsManager {
    static let shared = ParentalControlsManager()
    
    private let filterKey = "com.spotifyforteens.contentfilter"
    private let pinKey = "com.spotifyforteens.parentalpin"
    
    var currentFilter: ContentFilter {
        get {
            if let data = UserDefaults.standard.data(forKey: filterKey),
               let filter = try? JSONDecoder().decode(ContentFilter.self, forKey: data) {
                return filter
            }
            return .default
        }
        set {
            if let data = try? JSONEncoder().encode(newValue) {
                UserDefaults.standard.set(data, forKey: filterKey)
            }
        }
    }
    
    var isPinSet: Bool {
        return UserDefaults.standard.string(forKey: pinKey) != nil
    }
    
    func setPin(_ pin: String) {
        UserDefaults.standard.set(pin, forKey: pinKey)
    }
    
    func verifyPin(_ pin: String) -> Bool {
        return UserDefaults.standard.string(forKey: pinKey) == pin
    }
    
    func clearPin() {
        UserDefaults.standard.removeObject(forKey: pinKey)
    }
}

extension ContentFilter: Codable {
    enum CodingKeys: String, CodingKey {
        case maxRating, blockExplicitContent, blockedKeywords, allowedGenres
    }
}

File: lib/models/content_filter.dart
import 'track.dart';

/// Content rating levels
enum ContentRating {
  everyone(0),
  teen(1),
  mature(2),
  explicit(3);

  final int value;
  const ContentRating(this.value);
}

/// Content filter configuration
class ContentFilter {
  final ContentRating maxRating;
  final bool blockExplicitContent;
  final List<String> blockedKeywords;
  final List<String>? allowedGenres;

  const ContentFilter({
    required this.maxRating,
    required this.blockExplicitContent,
    required this.blockedKeywords,
    this.allowedGenres,
  });

  /// Default filter for teens (13+)
  static const defaultFilter = ContentFilter(
    maxRating: ContentRating.teen,
    blockExplicitContent: true,
    blockedKeywords: [],
    allowedGenres: null,
  );

  /// Strict filter for everyone
  static const strict = ContentFilter(
    maxRating: ContentRating.everyone,
    blockExplicitContent: true,
    blockedKeywords: ['drug', 'violence', 'explicit'],
    allowedGenres: ['pop', 'rock', 'indie', 'electronic', 'classical'],
  );

  /// Check if content is allowed based on filter rules
  bool isContentAllowed(Track track) {
    // Block explicit content if setting is enabled
    if (blockExplicitContent && track.isExplicit) {
      return false;
    }

    // Check for blocked keywords in track name
    final trackNameLower = track.name.toLowerCase();
    for (final keyword in blockedKeywords) {
      if (trackNameLower.contains(keyword.toLowerCase())) {
        return false;
      }
    }

    // Check for blocked keywords in artist names
    final artistNameLower = track.artistNames.toLowerCase();
    for (final keyword in blockedKeywords) {
      if (artistNameLower.contains(keyword.toLowerCase())) {
        return false;
      }
    }

    return true;
  }

  /// Serialize to JSON
  Map<String, dynamic> toJson() {
    return {
      'maxRating': maxRating.value,
      'blockExplicitContent': blockExplicitContent,
      'blockedKeywords': blockedKeywords,
      'allowedGenres': allowedGenres,
    };
  }

  /// Deserialize from JSON
  factory ContentFilter.fromJson(Map<String, dynamic> json) {
    return ContentFilter(
      maxRating: ContentRating.values[json['maxRating'] ?? 1],
      blockExplicitContent: json['blockExplicitContent'] ?? true,
      blockedKeywords: List<String>.from(json['blockedKeywords'] ?? []),
      allowedGenres: json['allowedGenres'] != null
          ? List<String>.from(json['allowedGenres'])
          : null,
    );
  }

  /// Create a copy with some fields updated
  ContentFilter copyWith({
    ContentRating? maxRating,
    bool? blockExplicitContent,
    List<String>? blockedKeywords,
    List<String>? allowedGenres,
  }) {
    return ContentFilter(
      maxRating: maxRating ?? this.maxRating,
      blockExplicitContent: blockExplicitContent ?? this.blockExplicitContent,
      blockedKeywords: blockedKeywords ?? this.blockedKeywords,
      allowedGenres: allowedGenres ?? this.allowedGenres,
    );
  }
}
